src/core/lib/channel/connected_channel.cc

```diff
@@ -203,39 +242,606 @@ static void connected_channel_get_channel_info(
+              gpr_log(GPR_INFO,
+                      "%sPollConnectedChannel: pushed message; requesting next",
+                      Activity::current()->DebugTag().c_str());
+            }
+            push_recv_message();
+          } else {
+            if (grpc_call_trace.enabled()) {
+              gpr_log(GPR_INFO,
+                      "%sPollConnectedChannel: pushed message and finished; "
+                      "marking closed",
+                      Activity::current()->DebugTag().c_str());
+            }
+            recv_message_state_ = Closed{};
+          }
+        } else {
+          if (grpc_call_trace.enabled()) {
+            gpr_log(GPR_INFO,
+                    "%sPollConnectedChannel: failed to push message; marking "
+                    "closed",
+                    Activity::current()->DebugTag().c_str());
+          }
+          recv_message_state_ = Closed{};
+        }
+      }
+    }
+    return Pending{};
+  }
+
+  void RecvInitialMetadataReady(grpc_error_handle error) {
+    GPR_ASSERT(error == GRPC_ERROR_NONE);
+    {
+      MutexLock lock(&mu_);
+      queued_initial_metadata_ = true;
+      initial_metadata_waker_.Wakeup();
+    }
+    Unref("initial_metadata_ready");
+  }
+
+  void RecvTrailingMetadataReady(grpc_error_handle error) {
+    GPR_ASSERT(error == GRPC_ERROR_NONE);
+    {
+      MutexLock lock(&mu_);
+      queued_trailing_metadata_ = true;
+      trailing_metadata_waker_.Wakeup();
+    }
+    Unref("trailing_metadata_ready");
+  }
+
+  void MetadataBatchDone(grpc_error_handle error) {
+    GPR_ASSERT(error == GRPC_ERROR_NONE);
+    Unref("metadata_batch_done");
+  }
+
+  void SendMessageBatchDone(grpc_error_handle error) {
+    {
+      MutexLock lock(&mu_);
+      if (error != GRPC_ERROR_NONE) {
+        send_message_state_ = Closed{};
+      }
+      if (!absl::holds_alternative<Closed>(send_message_state_)) {
+        send_message_state_ = Idle{};
+      }
+      send_message_waker_.Wakeup();
+    }
+    Unref("send_message");
+  }
+
+  void RecvMessageBatchDone(grpc_error_handle error) {
+    {
+      MutexLock lock(&mu_);
+      if (error != GRPC_ERROR_NONE) {
+        if (grpc_call_trace.enabled()) {
+          gpr_log(GPR_INFO, "%sRecvMessageBatchDone: error=%s",
+                  recv_message_waker_.ActivityDebugTag().c_str(),
+                  grpc_error_std_string(error).c_str());
+        }
+      } else if (absl::holds_alternative<Closed>(recv_message_state_)) {
+        if (grpc_call_trace.enabled()) {
+          gpr_log(GPR_INFO, "%sRecvMessageBatchDone: already closed, ignoring",
+                  recv_message_waker_.ActivityDebugTag().c_str());
+        }
+      } else {
+        auto pending =
+            absl::get_if<PendingReceiveMessage>(&recv_message_state_);
+        GPR_ASSERT(pending->received == false);
+        pending->received = true;
+      }
+      recv_message_waker_.Wakeup();
+    }
+    Unref("recv_message");
+  }
+
+  void Push() {
```

Comment:
Please add a comment here saying that this will be called outside the activity to push activity into the transport.

Reply:
done